Update cloud docs for the new hybrid cicd workflow (#14199)
dagster-io/dagster: docs/content/dagster-cloud/getting-started.mdx

3. In your `dagster-cloud-deploy.yml`, uncomment the step that is relevant to your Docker container registry. For example, if using DockerHub, you'd uncomment the `DockerHub` step in these files. Ensure you have set up the relevant secrets for building and uploading your Docker images.

After making the above changes and commiting the workflow file, the CI process should be triggered to deploy your GitHub repository to Dagster Cloud. During the deployment, the agent will attempt to load your code and update the metadata in Dagster Cloud. Once finished, you should see the GitHub Action complete successfully and also be able to see the code location under the "Deployment" tag in Dagster Cloud.

---

When finished, [continue to the next step](#step-4-set-up-environment-variables-and-secrets).

</TabItem>
<TabItem name="Other CI/CD provider">

For continuous integration using a CI/CD provider other than GitHub, your system should use the `dagster-cloud ci` subcommand to deploy code locations to Dagster Cloud.

> Ensure that you have created a `dagster_cloud.yaml` file as described in \[the quickstart]. For detailed documentation about this file see the [dagster_cloud.yaml reference](/dagster-cloud/managing-deployments/dagster-cloud-yaml). The following steps are typically implemented in the CI/CD workflow:

Author: Added a link to the quickstart repo.

1. Set the build environment variables:

   - `DAGSTER_CLOUD_ORGANIZATION` - \[required] the name of your organization in Dagster Cloud.
   - `DAGSTER_CLOUD_API_TOKEN` - \[required] a token created in Dagster Cloud. This is a sensitive value and should be stored as a CI/CD secret, if possible.
   - `DAGSTER_BUILD_STATEDIR` - \[required] path to a blank or non-existent temporary directory on the build machine. This directory is used to store local state during the build.

2. Run the configuration check:

   ```shell
   dagster-cloud ci check --project-dir=.
   ```

   This is an optional step but useful to validate the contents of your `dagster_cloud.yaml` and connection to Dagster Cloud.

3. Initialize the build session:

   ```shell
   dagster-cloud ci init --project-dir=.
   ```

   This reads the `dagster_cloud.yaml` configuration and initializes the `DAGSTER_BUILD_STATEDIR`.

4. Build and upload Docker images for your code locations.

>    The Docker image should contain a Python environment with `dagster`, `dagster-cloud` and your code. For reference, see the [example Dockerfile](https://github.com/dagster-io/dagster-cloud-hybrid-quickstart/blob/main/Dockerfile) in our quickstart repository. The example uses `pip install .` to install the code including the dependencies specified in [setup.py](https://github.com/dagster-io/dagster-cloud-hybrid-quickstart/blob/main/setup.py).

Reviewer: ```suggestion
   The Docker image should contain a Python environment with `dagster`, `dagster-cloud`, and your code. For reference, see the [example Dockerfile](https://github.com/dagster-io/dagster-cloud-hybrid-quickstart/blob/main/Dockerfile) in our quickstart repository. The example uses `pip install .` to install the code including the dependencies specified in [`setup.py`](https://github.com/dagster-io/dagster-cloud-hybrid-quickstart/blob/main/setup.py).
```

- Please use serial/oxford commas
- Adding code formatting to `setup.py`
- Is the period in `pip install .` correct? (I'm not familiar with this, so I'm asking to make sure it's not a typo)

Author: Yes the period in `pip install .` is required.

   It is a good idea to use a unique image tag for each Docker build. You can build one image per code location or a shared image for multiple code locations. As an example image tag, you can use the git commit SHA:

   ```shell
   export IMAGE_TAG=`git log --format=format:%H -n 1`
   ```

   Use this tag to build and upload your Docker image, for example:

   ```shell
   docker build . -t ghcr.io/org/dagster-cloud-image:$IMAGE_TAG
   docker push ghcr.io/org/dagster-cloud-image:$IMAGE_TAG
   ```

   The upload step is specific to your Docker container registry and will require authentication. The only requirement is that the registry you upload to must match the registry specified in `dagster_cloud.yaml`.

5. Update the build session with the Docker image tag. For each code location you want to deploy, run the following command passing the `IMAGE_TAG` used in the previous step:

   ```shell
   dagster-cloud ci set-build-output --location-name=code-location-a --image-tag=IMAGE_TAG
   ```

   This command does not deploy the code location but just updates the local state in `DAGSTER_BUILD_STATEDIR`.

6. Deploy to Dagster Cloud:

   ```shell
   dagster-cloud ci deploy
   ```

>    This command updates the code locations in Dagster Cloud. Once this finishes successfully, you should be able to see the code locations under the "Deployments" tab in Dagster Cloud.

Reviewer: ```suggestion
   This command updates the code locations in Dagster Cloud. Once this finishes successfully, you should be able to see the code locations under the **Deployments** tab in Dagster Cloud.
```

UI components should use bold formatting.

</TabItem>
<TabItem name="Other CI/CD provider">

For continuous integration using a CI/CD provider other than GitHub, your system should use the `dagster-cloud ci` subcommand to deploy code locations to Dagster Cloud.

Ensure that you have created a `dagster_cloud.yaml` file as described in [the quickstart](https://github.com/dagster-io/dagster-cloud-hybrid-quickstart/). For detailed documentation about this file see the [`dagster_cloud.yaml` reference](/dagster-cloud/managing-deployments/dagster-cloud-yaml). The following steps are typically implemented in the CI/CD workflow:

1. Set the build environment variables. Note that all variables are required:

   - `DAGSTER_CLOUD_ORGANIZATION` - The name of your organization in Dagster Cloud.
   - `DAGSTER_CLOUD_API_TOKEN` - A Dagster Cloud API token. **Note**: This is a sensitive value and should be stored as a CI/CD secret, if possible.
   - `DAGSTER_BUILD_STATEDIR` - A path to a blank or non-existent temporary directory on the build machine. This directory is used to store local state during the build.
   dagster-cloud ci check --project-dir=.

>    ```

Reviewer: I think some formatting got trampled on here with the code block

1. In the repository, set the necessary GitHub action secrets:

   - `DAGSTER_CLOUD_API_TOKEN` - The Dagster Cloud agent token from the previous section. Refer to the [agent tokens documentation](/dagster-cloud/account/managing-user-agent-tokens#managing-agent-tokens) for more info.

   Refer to the [GitHub docs](https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-a-repository) for more info about GitHub Action secrets.

> 2. In `deploy-dagster-cloud.yml`, set the `DAGSTER_CLOUD_ORGANIZATION` environment variable to your dagster.cloud organization name.

Author: I agree. I think all of this info in the "getting started" page is a bit overwhelming. Many of the links point to later sections in the docs because you need to know about code locations and tokens before you can implement CI/CD.

I'm now thinking we should have a separate page dedicated to setting up CI/CD somewhere under the Cloud section: 
<img width="277" alt="image" src="https://github.com/dagster-io/dagster/assets/7066873/f0ed7a59-521b-4c86-913b-6b849f884aa2">

> 2. In `deploy-dagster-cloud.yml`, set the `DAGSTER_CLOUD_ORGANIZATION` environment variable to your dagster.cloud organization name.

Reviewer: We should. Branch Deployments currently serves the CI portion of this in the docs, but it's not immediately clear that that's what that term means.

Should CI/CD go under **Managing deployments**?

> 2. In `deploy-dagster-cloud.yml`, set the `DAGSTER_CLOUD_ORGANIZATION` environment variable to your dagster.cloud organization name.

Author: For now I've added a link to our quickstart's workflow file and also added some info about what happens when the workflow runs.

3. In `deploy-dagster-cloud.yml`, uncomment the step that is relevant to your Docker container registry. For example, if using DockerHub, you'd uncomment the `DockerHub` step in these files. Ensure you have set up the relevant secrets for building and uploading your Docker images.

> When finished setting up the repository, navigate back to Dagster Cloud and move on to the next step.

Author: Updated.

</TabItem>
<TabItem name="Other CI/CD provider">

For continuous integration using a CI/CD provider other than GitHub, your system should use the `dagster-cloud ci` subcommand to deploy code locations to Dagster Cloud.

> Ensure that you have created a `dagster_cloud.yaml` file as described in [Syncing the workspace](/dagster-cloud/managing-deployments/code-locations#managing-code-locations-using-the-dagster-cloud-cli#syncing-the-workspace). The following steps are typically implemented in the CI/CD workflow:

Author: Linked to our example quickstart's `dagster_cloud.yaml` file.

1. Set the build environment variables:

   - `DAGSTER_CLOUD_ORGANIZATION` - the name of your organization in Dagster Cloud.
   - `DAGSTER_CLOUD_API_TOKEN` - a token created in Dagster Cloud.

>    - `DAGSTER_BUILD_STATEDIR` - path to a blank or nonexistant temporary directory on the build machine. This directory is used to store local state during the build.

Reviewer: ```suggestion
   - `DAGSTER_BUILD_STATEDIR` - path to a blank or non-existent temporary directory on the build machine. This directory is used to store local state during the build.
```

>    - `DAGSTER_BUILD_STATEDIR` - path to a blank or nonexistant temporary directory on the build machine. This directory is used to store local state during the build.

Author: Added `[required]` tags next to each

> 2. Run the configuration check:
> 
>        dagster-cloud ci check --project-dir=.

Reviewer: ````suggestion
2. Run the configuration check:

   ```shell
   dagster-cloud ci check --project-dir=.
   ```
````

Make sure to put code snippets into fenced blocks and include a language. CSS isn't applied correctly without it

   This is an optional step but useful to validate the contents of your `dagster_cloud.yaml` and connection to Dagster Cloud.

> 3. Initialize the build session:
> 
>        dagster-cloud ci init --project-dir=.

Reviewer: ````suggestion
3. Initialize the build session:

   ```shell
   dagster-cloud ci init --project-dir=.
   ```
````

   This reads the `dagster_cloud.yaml` configuration and initializes the `DAGSTER_BUILD_STATEDIR`.

> 4. Build and upload Docker images for you code locations.

Reviewer: +1

> 4. Build and upload Docker images for you code locations.

Reviewer: ```suggestion
4. Build and upload Docker images for your code locations.
```

> 4. Build and upload Docker images for you code locations.

Author: Linked to our quickstart's Dockerfile as an example.

   This step is specific to your Docker container registry and Docker build process. The only requirement is that the registry you upload to must match the registry specified in `dagster_cloud.yaml`. It is a good idea to use a unique image tag for each build.

   You can build one image per code location or a shared image for multiple code locations.

> 5. Update the build session with the Docker image tag. For each code location you want to deploy, run the following command passing the `IMAGE_TAG` used in the previous step:

Author: Fixed - added more detail and an example `IMAGE_TAG` to the previous section.

> 5. Update the build session with the Docker image tag. For each code location you want to deploy, run the following command passing the `IMAGE_TAG` used in the previous step:
> 
>        dagster-cloud ci set-build-output --location-name=code-location-a --image-tag=IMAGE_TAG

Reviewer: ````suggestion
5. Update the build session with the Docker image tag. For each code location you want to deploy, run the following command passing the `IMAGE_TAG` used in the previous step:

   ```shell
   dagster-cloud ci set-build-output --location-name=code-location-a --image-tag=IMAGE_TAG
   ```
````

   This command does not deploy the code location but just updates the local state in `DAGSTER_BUILD_STATEDIR`.

> 6. Deploy to Dagster Cloud:
> 
>        dagster-cloud ci deploy

Reviewer: ````suggestion
6. Deploy to Dagster Cloud:

   ```shell
   dagster-cloud ci deploy
   ```
````
